driver/src/price_finding/naive_solver.rs

```diff
@@ -180,32 +208,96 @@ fn order_with_buffer_for_fee(order: &Order, fee: &Option<Fee>) -> Order {
     }
 }
 
+/// Normalizes a price base on the pre-normalized fee price.
+fn normalize_price(price: u128, pre_normalized_fee_price: u128) -> Option<u128> {
+    // upcast to u256 to avoid overflows
+    checked_u256_to_u128(
+        (u128_to_u256(price) * u128_to_u256(BASE_PRICE))
+            .ceiled_div(u128_to_u256(pre_normalized_fee_price)),
+    )
+}
+
+/// Calculate the executed sell amount from the fee, executed buy amount, and
+/// the buy and sell prices of the traded tokens.
+fn executed_sell_amount(fee: &Fee, exec_buy_amt: u128, buy_price: u128, sell_price: u128) -> u128 {
+    let fee_denominator = (1.0 / fee.ratio) as u128;
+    u256_to_u128(
+        (((u128_to_u256(exec_buy_amt) * u128_to_u256(buy_price))
+            / u128_to_u256(fee_denominator - 1))
+            * u128_to_u256(fee_denominator))
+            / u128_to_u256(sell_price),
+    )
+}
+
+/// Calculate the executed buy amount from the fee, executed sell amount, and
+/// the buy and sell prices of the traded tokens. This function returns a `None`
+/// if no value can be found such that `executed_sell_amount(result, pb, ps) ==
+/// xs`. This function acts as an inverse to `executed_sell_amount`.
+fn executed_buy_amount(
+    fee: &Fee,
+    exec_sell_amt: u128,
+    buy_price: u128,
+    sell_price: u128,
+) -> Option<u128> {
+    let fee_denominator = (1.0 / fee.ratio) as u128;
+    let exec_buy_amt = u256_to_u128(
+        (((u128_to_u256(exec_sell_amt) * u128_to_u256(sell_price))
+            / u128_to_u256(fee_denominator))
+            * u128_to_u256(fee_denominator - 1))
+            / u128_to_u256(buy_price),
+    );
+
+    // we need to account for rounding errors here, since this function is
+    // essentially an inverse of `executed_sell_amount`; so find the maximum
+    // error that `exec_buy_amt` can have and find a value that works
+
+    macro_rules! return_if_correct {
+        ($proposed_exec_buy_amt:expr) => {
+            if exec_sell_amt
+                == executed_sell_amount(fee, $proposed_exec_buy_amt, buy_price, sell_price)
+            {
+                return Some($proposed_exec_buy_amt);
+            }
+        };
+    }
+
+    return_if_correct!(exec_buy_amt);
+
+    let maximum_error = (sell_price / buy_price) + 1;
+    for error in 1..=maximum_error {
```

Comment:
What's the bound of this loop? Do we have to be afraid to run it a billion times?

Reply:
My understanding of the bound here was... `999_000` This can also be found (hardcoded) as a constant in the `dex-contracts/test/resources/math.js` as

```
const ERROR_EPSILON = new BN(999000)
```

Reply:
It potentially could for small buy prices and high sell prices. I guess we could put an upper bound on this. Also, let me try to formalize a bit what this represents so it feels a bit less magic. Essentially it is supposed to be the maximum error that the executed buy amount could have so that `x == executed_sell_amount(executed_buy_amount(x, buy_price, sell_price), buy_price, sell_price)`.

Reply:
We need this since the naive solver relies on fixing the executed buy amounts for orders that are selling the fee token and fixing the executed sell amounts for orders that are buying the fee token (since the rounding errors must be on the fee token).

Reply:
Hmm, now that I think about it, I might have made a mistake here.

Reply:
I think the `maximum_error == (fee_denominator / buy_price) + 1` which, because of price restrictions, is just 1.

It might also be `((fee_denominator / buy_price) + 1) + (((fee_denominator - 1) / sell_price) + 1) == 2` to account for errors in the `executed_sell_amount` calculation as well. Also I would expect the error to always be in the same direction so we might not have to check in both directions.

 I'm coming into the office in an hour or so, can someone help me look at this?

Reply:
@bh2smith That error epsilon is for something else I think. If I am not mistaken it is the maximum error for buy and sell order amounts so that they match an exact price. Something like:
```
order      price     order
    [              |              ]
-----                          -----
ERROR_EPSILON
```

Reply:
Turns out we needed to do a ceiled div. There are still values of executed sell amount that have no inverse in integer math. A future improvement would be to have the naive solver that deals with this by adjusting amounts to work.

Note that if the price of the non-fee token is not smaller than the base price, then the inverse exists and can be computed.